src/plugins/offboard/include/plugins/offboard/offboard.h

```diff
@@ -131,6 +131,29 @@ class Offboard : public PluginBase {
                                percent). */
     };
 
+    /**
+     * @brief Type for actuator control command.
+     * controls members should be normed to -1..+1 where 0 is neutral position.
+     * Throttle for single rotation direction motors is 0..1, negative range for reverse direction.
+     *
+     * One group support eight controls.
+     *
+     * In PX4 v1.9.0 Only first four Control Groups are supported
+     * (https://github.com/PX4/Firmware/blob/v1.9.0/src/modules/mavlink/mavlink_receiver.cpp#L980).
+     */
+    struct ActuatorControl {
+        float controls[16]; /**< @brief Actuator controls array.
+                               First 8 will go to control group 0, the following 8 to control group 1.
+                               If you want to send less than 16 controls you should to set first non-specified
+                               control to NAN.
+                               If index of first NAN value will be less or equal than 8 then only one
+                               message for Group 0 will be sent. All controls from first NAN value
+                               to 7 in Group 0 will be set to zero.
+                               If index of first NAN value will be great than 8 then two messages (to
+                               Group 0 and Group 1) will be sent. In this case all controls from first NAN value
+                               to 7 in Group 1 will be set to zero. */
```

Comment:
Perhaps this description should be more understandable for native English speakers :)

Reply:
:smile: what about:

```
Up to 16 actuator controls can be set. To ignore an output, set it to NaN.
The first 8 actuator controls internally map to control group 0, the latter 8 actuator controls map to control group 1.
Depending on what controls are set (instead of NaN) 1 or 2 MAVLink messages are actually sent.
```

Given that description is a bit awkward, would it be clearer, like that:

```
struct ActuatorControl {
    struct Group {
        float controls[8];
    };
    Group groups[2];
}
```

(Or is that almost what you had suggested in the beginning? Sorry for the long discussion)

Reply:
All ok, I understand that you just want to construct good solution :)


> ```
> struct ActuatorControl {
>     struct Group {
>         float controls[8];
>     };
>     Group groups[2];
> }
> ```

This is the possible solution, but In this case, in my assumption we need to change [ActuatorControl structure in Proto](https://github.com/mavlink/MAVSDK-Proto/pull/88/files#diff-d071a0f40909dedf07ff7e03789cb56aR72) too due to API "consistency". It can be like that:

```
message ControlGroup {
  repeated float controls = 1;
}
message ActuatorControl {
  repeated ControlGroup control_group = 1;
}
```

But this can be too comlpicated?

-------

If you just don't want to have field like `group_number` in data types then I have this proposal:

```
struct ActuatorControl {
    struct Group {
        float controls[8];
    };
    union {
        Group groups[2];
        float controls[16];
    }
}
```
In this case [current](https://github.com/mavlink/MAVSDK-Proto/pull/88/files#diff-d071a0f40909dedf07ff7e03789cb56aR72) protobuf message can be used.
Also description of usage can be like that:

```suggestion
Up to 16 actuator controls can be set. To ignore an **output group**, set it to NaN. In case of **all** controls in the group is NaN than this group will not be sent (otherwise all NaN controls will sent as zero).
The first 8 actuator controls internally map to control group 0, the latter 8 actuator controls map to control group 1.
Depending on what controls are set (instead of NaN) 1 or 2 MAVLink messages are actually sent. */
```

What do you think about it?

Reply:
> ```
> message ControlGroup {
>  repeated float controls = 1;
>}
>message ActuatorControl {
>  repeated ControlGroup control_group = 1;
>}
>```
>
> But this will be too comlpicated?

I'd say that's fine.

Reply:
Okay, I will implement it.
In this case the `union` is not required.

Reply:
Done